ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/peervalidation/RequiredBlocksPeerValidator.java

```diff
@@ -0,0 +1,132 @@
+ *  * the License. You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ *  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  * specific language governing permissions and limitations under the License.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.eth.peervalidation;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.eth.manager.task.AbstractPeerTask;
+import org.hyperledger.besu.ethereum.eth.manager.task.GetHeadersFromPeerByNumberTask;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class RequiredBlocksPeerValidator implements PeerValidator {
+  private static final Logger LOG = LogManager.getLogger();
+  private static long DEFAULT_CHAIN_HEIGHT_ESTIMATION_BUFFER = 10L;
+
+  private final ProtocolSchedule<?> protocolSchedule;
+  private final MetricsSystem metricsSystem;
+
+  private final long blockNumber;
+  private final Hash hash;
+  // Wait for peer's chainhead to advance some distance beyond blockNumber before validating
+  private final long chainHeightEstimationBuffer;
+
+  RequiredBlocksPeerValidator(
+      final ProtocolSchedule<?> protocolSchedule,
+      final MetricsSystem metricsSystem,
+      final long blockNumber,
+      final Hash hash,
+      final long chainHeightEstimationBuffer) {
+    checkArgument(chainHeightEstimationBuffer >= 0);
+    this.protocolSchedule = protocolSchedule;
+    this.metricsSystem = metricsSystem;
+    this.blockNumber = blockNumber;
+    this.hash = hash;
+    this.chainHeightEstimationBuffer = chainHeightEstimationBuffer;
+  }
+
+  public RequiredBlocksPeerValidator(
+      final ProtocolSchedule<?> protocolSchedule,
+      final MetricsSystem metricsSystem,
+      final long blockNumber,
+      final Hash hash) {
+    this(
+        protocolSchedule, metricsSystem, blockNumber, hash, DEFAULT_CHAIN_HEIGHT_ESTIMATION_BUFFER);
+  }
+
+  @Override
+  public CompletableFuture<Boolean> validatePeer(
+      final EthContext ethContext, final EthPeer ethPeer) {
+    final AbstractPeerTask<List<BlockHeader>> getHeaderTask =
+        GetHeadersFromPeerByNumberTask.forSingleNumber(
```

Comment:
You probably want to give them a few chances with: `RetryingGetHeaderFromPeerByNumberTask`

Reply:
If we go that path we lose our timeout, it extends from a different super class that doesn't have timeouts support, exchanging it for retries.  I view it as mostly equivalent, but since this is a first request to a new peer I think timeouts are more useful.

Reply:
It also gets in the way of refactoring out common code if we change it.